extensions/amp-bind/0.1/bind-impl.js

```diff
@@ -119,50 +111,93 @@ export class Bind {
     Object.assign(this.scope_, state);
 
     if (!opt_skipDigest) {
-      this.digest_();
+      // If scan hasn't completed yet, set `digestQueuedAfterScan_`.
+      if (this.boundElements_) {
+        this.digest_();
+      } else {
+        this.digestQueuedAfterScan_ = true;
+      }
     }
   }
 
   /**
+   * Scans the ampdoc for bindings and creates the expression evaluator.
+   * @private
+   */
+  initialize_() {
+    this.scanPromise_ = this.scanForBindings_(this.ampdoc.getBody());
+    this.scanPromise_.then(results => {
+      const {boundElements, evaluatees} = results;
+
+      this.boundElements_ = boundElements;
+      this.evaluator_ = new BindEvaluator(evaluatees);
+
+      // Trigger verify-only digest in development.
+      const development = getMode().development;
+      if (development || this.digestQueuedAfterScan_) {
+        this.digest_(/* opt_verifyOnly */ development);
+      }
+    });
+  }
+
+  /**
    * Scans `body` for attributes that conform to bind syntax and returns
    * a tuple containing bound elements and binding data for the evaluator.
    * @param {!Element} body
-   * @return {{
-   *   boundElements: !Array<BoundElementDef>,
-   *   evaluatees: !Array<./bind-evaluator.EvaluateeDef>,
-   * }}
+   * @return {
+   *   !Promise<{
+   *     boundElements: !Array<BoundElementDef>,
+   *     evaluatees: !Array<./bind-evaluator.EvaluateeDef>
+   *   }>
+   * }
    * @private
    */
   scanForBindings_(body) {
-    // TODO(choumx): Chunk if taking too long in a single frame.
-
     /** @type {!Array<BoundElementDef>} */
     const boundElements = [];
     /** @type {!Array<./bind-evaluator.EvaluateeDef>} */
     const evaluatees = [];
 
+    // TODO(choumx): Use TreeWalker if this is too slow.
     const elements = body.getElementsByTagName('*');
-    for (let i = 0; i < elements.length; i++) {
-      const element = elements[i];
-      const attributes = element.attributes;
-
-      const bindings = [];
-      for (let j = 0; j < attributes.length; j++) {
-        const binding = this.bindingForAttribute_(attributes[j], element);
-        if (binding) {
-          bindings.push(binding);
-          evaluatees.push({
-            tagName: element.tagName,
-            property: binding.property,
-            expressionString: binding.expressionString,
-          });
+
+    // Helper function for scanning a slice of elements.
+    const scanFromTo = (start, end) => {
+      for (let i = start; i < Math.min(end, elements.length); i++) {
```

Comment:
Don't remember 100%, but I believe calling `elements.length` is here would force some significant amount of work. It might be better to iterate until we arrive at `elements[++i] === undefined`.